src/TopOptEMFocus.jl

```diff
@@ -0,0 +1,550 @@
+        ∫(a_f(r, u, v))fem_params.dΩ_d + ∫(v * u)fem_params.dΩ_d, ∫(v * ph)fem_params.dΩ_d
+      end
+    pfh = solve(op)
+    return get_free_dof_values(pfh)
+end
+
+function Threshold(pfh; β, η)
+    return ((tanh(β * η) + tanh(β * (pfh - η))) / (tanh(β * η) + tanh(β * (1.0 - η))))         
+end
+
+
+# ## Weak form
+# 
+# We notice that the design parameters only affect the weak form in the design domain and the PML does not affect the design domain, we can then make things simpler by dividing the weak form to a base integral that contains the whole computation cell and an additional integral on the design domain. We also make a LU factorization on the final Maxwell operator matrix $A$ since it will only be used to solve for linear equations. 
+# 
+
+using LinearAlgebra
+ξd(p, n_air, n_metal)= 1 / (n_air + (n_metal - n_air) * p)^2 - 1 / n_air^2 # in the design region
+
+a_base(u, v; phys_params) = (1 / phys_params.n_air^2) * ((∇ .* (Λ(phys_params) * v)) ⊙ (Λ(phys_params) .* ∇(u))) - (phys_params.k^2 * phys_params.μ * (v * u))
+
+a_design(u, v, pth; phys_params) = ((p -> ξd(p, phys_params.n_air, phys_params.n_metal)) ∘ pth) * (∇(v) ⊙ ∇(u))
+
+function MatrixA(pth; phys_params, fem_params)
+    A_mat = assemble_matrix(fem_params.U, fem_params.V) do u, v
+        ∫(a_base(u, v; phys_params))fem_params.dΩ + ∫(a_design(u, v, pth; phys_params))fem_params.dΩ_d
+    end
+    return lu(A_mat)
+end
+
+# ## Solve for plane wave incident
+# 
+# The plane wave source `b_vec` can be simply assembled by a uniform integral over the source line, and the magnetic field vector `u_vec` can then be solved by a simple linear equation
+# 
+
+p0 = zeros(fem_params.np)  # Here we make p=0 everywhere just for illustration purpose
+pf_vec = Filter(p0;r, fem_params)
+pfh = FEFunction(fem_params.Pf, pf_vec)
+pth = (pf -> Threshold(pf; β, η)) ∘ pfh
+A_mat = MatrixA(pth; phys_params, fem_params) 
+b_vec = assemble_vector(v->(∫(v)fem_params.dΓ_s), fem_params.V)
+u_vec = A_mat \ b_vec
+uh = FEFunction(fem_params.U, u_vec)
+
+# ## Objective
+# 
+# The problem is maximizing the electric field intensity at the center. Recall that the electric field can be retrieved from the magnetic field by 
+#
+# ```math
+# \mathbf{E}(\mathbf{x})=\frac{\mathrm{i}}{\omega\varepsilon(\mathbf{x})}\nabla\times\mathbf{H}(\mathbf{x}),
+# ``` 
+# and our objective is the field intensity at center $\vert\mathbf{E}(\mathbf{x}_0)\vert^2$. 
+# 
+# In the 2D formulation, this objective can be simplified to  
+# ```math
+# g = \int \vert\nabla H\vert^2\delta(x-x_0)\mathrm{d}\Omega = u^\dagger Ou,
+# ```
+# where $u$ is the magnetic field vector and  
+# ```math
+# O = \int (\nabla \hat{v}\cdot\nabla\hat{u})\delta(x-x_0)\mathrm{d}\Omega,
+# ```
+# with $\hat{v}$ and $\hat{u}$ are the finite element basis functions. 
+# 
+# In practice, the delta function can be approximated by a concentrated Gaussian function. Note that we use `dΩ_c` here in order to reduce computation costs. 
```

Comment:
```suggestion
# In practice, the delta function can be approximated narrow Gaussian peak. Note that we use `dΩ_c` here in order to reduce the computational cost of the quadrature (to avoid integrating in regions where the Gaussian weight is negligible). 
```